### ISSUE TITLE
[Feature Request] Build-in IK / Inverse Kinematic Solver

### ISSUE THREAD

**Issue by Kenji245** [xenko3d/xenko#103]
beside to handle the character foot position, maybe this will be useful for 3rd person shooter game.

**Comment by Vega-KH** [comment]
A first step in this direction would be to expose bones in the API.  This would allow developers to at least set bone positions and rotation manually and could be used to script a basic IK solver.

**Comment by Kenji245** [comment]
yes, well, i only need to rotate a single bone manually, to rotate the spine bone according to mouse y axis, wasn't really IK though, but if xenko have IK that would be cool too.

**Comment by Jaygem** [comment]
Is there a way to work on the API to expose the bones ?

**Comment by Eideren** [comment]
I haven't used it but [this](https://github.com/xenko3d/xenko/blob/243f7a33c9020b39682ef7d336757eb2027a8b15/sources/engine/Xenko.Engine/Engine/ModelComponent.cs#L108) seems like what you are looking for.

**Comment by Jaygem** [comment]
Thanks @Eideren i think i found something with

**Comment by Eideren** [comment]
@Jaygem use `NodeTransformations`'s `Transform` instead, works perfectly on my end.

**Comment by Jaygem** [comment]
@Eideren I'm sorry it doesn't work for me :c 
How do you do it?

**Comment by Eideren** [comment]
@Jaygem 
```
var Skeleton = Entity.Get<ModelComponent>().Skeleton;
for( int i = 0; i < Skeleton.NodeTransformations.Length; i++ )
{
	Skeleton.NodeTransformations[ i ].Transform.Position = new Vector3( 0f, 0f, 0f );
}
```
Put this inside your update function, it should place all of your bones at the origin.

**Comment by anan1213095357** [comment]
When can I receive support? A game engine lacking IK and animation state machines will greatly increase the entry threshold

**Comment by anan1213095357** [comment]
> @JAYGEM
> 
> ```
> var Skeleton = Entity.Get<ModelComponent>().Skeleton;
> for( int i = 0; i < Skeleton.NodeTransformations.Length; i++ )
> {
> 	Skeleton.NodeTransformations[ i ].Transform.Position = new Vector3( 0f, 0f, 0f );
> }
> ```
> 
> Put this inside your update function, it should place all of your bones at the origin.

When can I receive support? A game engine lacking IK and animation state machines will greatly increase the entry threshold

**Comment by Eideren** [comment]
@anan1213095357 Stride is not a commercial engine, every person contributing does so on their free time. If you want a specific feature to get implemented in the engine your best bet is to do it yourself or pay someone to do it for you

**Comment by anan1213095357** [comment]
> 
If my technical level could be achieved, I wouldn't be asking questions here. Because this ik is considered a necessary content for a mature game engine. So I want to see if everyone has the will to achieve it

**Comment by anan1213095357** [comment]
> @anan1213095357Stride 不是商业引擎，每个贡献者都是在空闲时间这样做的。如果您希望在引擎中实现特定功能，您最好的选择是自己做或花钱请人为您做

If my technical level could be achieved, I wouldn't be asking questions here. Because this ik is considered a necessary content for a mature game engine. So I want to see if everyone has the will to achieve it

**Comment by anan1213095357** [comment]
> @anan1213095357Stride 不是商业引擎，每个贡献者都是在空闲时间这样做的。如果您希望在引擎中实现特定功能，您最好的选择是自己做或花钱请人为您做



> @anan1213095357 Stride is not a commercial engine, every person contributing does so on their free time. If you want a specific feature to get implemented in the engine your best bet is to do it yourself or pay someone to do it for you

If my technical level could be achieved, I wouldn't be asking questions here. Because this ik is considered a necessary content for a mature game engine. So I want to see if everyone has the will to achieve it

**Comment by anan1213095357** [comment]
> @anan1213095357 Stride is not a commercial engine, every person contributing does so on their free time. If you want a specific feature to get implemented in the engine your best bet is to do it yourself or pay someone to do it for you

If my technical level could be achieved, I wouldn't be asking questions here. Because this ik is considered a necessary content for a mature game engine. So I want to see if everyone has the will to achieve it

**Comment by Perksey** [comment]
@anan1213095357 looks like you’ll want to find a “more mature” engine then. Stride has lots of users (that do think it’s mature enough for them) and they already have lots of needs that are being worked on, so unless anyone has particular passion for this specific feature (there has been no indication as of yet) then it isn’t going to be worked on in the near future. This isn’t me speaking as a Stride dev (because I’m not), this is just the way open source works. Open source and the volunteers therein does not mean the world is going to materialise around you.

**Comment by anan1213095357** [comment]
> @anan1213095357 looks like you’ll want to find a “more mature” engine then. Stride has lots of users (that do think it’s mature enough for them) and they already have lots of needs that are being worked on, so unless anyone has particular passion for this specific feature (there has been no indication as of yet) then it isn’t going to be worked on in the near future. This isn’t me speaking as a Stride dev (because I’m not), this is just the way open source works. Open source and the volunteers therein does not mean the world is going to materialise around you.

"I believe the future of a thriving user community for a game engine still greatly involves newcomers like me. The reason I'm drawn to this game engine is because, compared to others, it offers simplicity and a low learning curve, which is perfect for beginners. However, I do face challenges when certain tools necessary for my projects are not available in the game engine, making it impossible for newcomers like me to achieve certain goals. My hope is that this game engine continues to improve and grow. That's why I brought up this point."

**Comment by ExtraMedicated** [comment]
I was relying heavily on the RootMotion Final IK package in my Unity projects. I was hoping there was some alternative here, but porting that over seems like a lot more than I'm personally capable of.
